lib/streamlit/runtime/caching/cache_errors.py

```diff
@@ -140,10 +165,27 @@ def __init__(
 
         super().__init__(msg)
 
-    @staticmethod
-    def _get_cached_func_name_md(func: types.FunctionType) -> str:
-        """Get markdown representation of the function name."""
-        if hasattr(func, "__name__"):
-            return "`%s()`" % func.__name__
-        else:
-            return "a cached function"
+
+class UnserializableReturnValueError(MarkdownFormattedException):
+    def __init__(self, func: types.FunctionType, return_value: types.FunctionType):
+        MarkdownFormattedException.__init__(
+            self,
+            f"""
+            Cannot serialize the return value (of type {get_return_value_type(return_value)}) in {_get_cached_func_name_md(func)}.  
+            `st.experimental_memo` uses [pickle](https://docs.python.org/3/library/pickle.html) to 
+            serialize the function’s return value and safely store it in the cache without mutating the original object. Please convert the return value to a pickle-serializable type.  
+            If you want to cache unserializable objects such as database connections or Tensorflow 
+            sessions, use `st.experimental_singleton` instead (see [our docs](https://docs.streamlit.io/library/advanced-features/experimental-cache-primitives) for differences).""",
+        )
+
+
+class UnevaluatedDataFrameError(MarkdownFormattedException):
```

Comment:
I don't think it makes sense to introduce a new exception type here since this is probably a pretty uncommon error. Just use whatever we use elsewhere for general error messages (I think StreamlitAPIException?). Not sure though, can you ask one of the more experienced devs which error type we usually use in this situation?

Reply:
Sure 👍 I've removed extra exception, 🤔 , but I've used `MarkdownFormattedException`, because one can follow stack trace to see which lines of code coused it, which can be helpful. But since you told me to ask more experienced dev I'll write to @LukasMasuch. Lukas, can you please review it?

Reply:
@jrieke however to avoid exception being shown as `MarkdownFormattedException` I've reverted back inline class `UnevaluatedDataFrameError`, let me know if that's ok, if not I will use `StreamlitAPIException`, however personally I don't really like hiding stacktrace from user

Reply:
I'm actually more inclined to add more specific exception types instead of using generic ones. The reason is that we can actually distinguish those with the new metrics tracking. And having some insights into this is probably worth it to define some more fine-grained exceptions. However, we should make sure to reuse this exception type for all semantically similar situations.

Reply:
@sfc-gh-tszerszen Regarding `StreamlitAPIException` vs `MarkdownFormattedException`: Can you post a quick example stack trace for both types here?

Reply:
Here is current version stacktrace:
```
File "/Users/tszerszen/Documents/streamlit/lib/streamlit/runtime/scriptrunner/script_runner.py", line 563, in _run_script
    exec(code, module.__dict__)
File "/Users/tszerszen/Documents/streamlit/examples/snowpark/stsnowpark.py", line 29, in <module>
    df = load_data()
File "/Users/tszerszen/Documents/streamlit/lib/streamlit/runtime/caching/cache_utils.py", line 264, in wrapper
    return get_or_create_cached_value()
File "/Users/tszerszen/Documents/streamlit/lib/streamlit/runtime/caching/cache_utils.py", line 256, in get_or_create_cached_value
    raise UnserializableReturnValueError(
```

And here is `StreamlitAPIException` stacktrace:

```
File "/Users/tszerszen/Documents/streamlit/examples/snowpark/stsnowpark.py", line 29, in <module>
    df = load_data()
```

Reply:
I'm slightly more in favour of `StreamlitAPIException` here (`UnevaluatedDataFrameError(StreamlitAPIException)`) but no strong opinion. The reason are 1) it's more consistent with other API usage related exceptions in Streamlit 2) For this error, we actually 100% know the reason and solution. I don't think there is any additional insights the user can get from the Stacktrace with the Streamlit internals in this case. 

We plan to improve our general error handling next year, so this might be something to revisit if we should - in general - show more stacktrace for API errors.

Reply:
> I'm slightly more in favour of StreamlitAPIException here
Sure 👍 It's a detail I guess and a matter of personal preference, I've updated it to use `StreamlitAPIException`. 

Are we good to merge it?